tests/test_keyword_matcher.py

```diff
@@ -0,0 +1,221 @@
+        if len(matches) > 1:
+            assert matches[0].confidence >= matches[-1].confidence
+
+        # Check that matched_keywords contains the correct keywords
+        for match in matches:
+            assert all(kw in keywords for kw in match.matched_keywords)
+
+    def test_match_by_substring_special_characters(self, keyword_matcher):
+        """Test substring matching with special characters in keywords."""
+        keywords = ["sequence-alignment", "rna/dna"]
+        matches = keyword_matcher.match_by_substring(keywords)
+
+        # Should handle gracefully (may or may not match depending on preprocessing)
+        assert isinstance(matches, list)
+
+    # Tests for Mode 2: List Embeddings
+
+    def test_match_by_list_embeddings_initialization(self, keyword_matcher):
+        """Test that embeddings are initialized when needed."""
+        keywords = ["alignment"]
+        matches = keyword_matcher.match_by_list_embeddings(keywords)
+
+        # Embeddings should be created during first call
+        assert keyword_matcher.embedding_model is not None
+        assert len(keyword_matcher.term_embeddings) > 0
+        assert isinstance(matches, list)
+
+    def test_match_by_list_embeddings_single_keyword(self, keyword_matcher):
+        """Test list embeddings matching with a single keyword."""
+        keywords = ["alignment"]
+        matches = keyword_matcher.match_by_list_embeddings(keywords, threshold=0.3)
+
+        assert isinstance(matches, list)
+        assert len(matches) > 0
+
+        # Check match structure
+        for match in matches:
+            assert isinstance(match, KeywordMatch)
+            assert match.concept_uri
+            assert match.concept_label
+            assert match.confidence >= 0.3
+            assert match.match_type == "list_embeddings"
+            assert len(match.matched_keywords) == 1
+            assert match.matched_keywords[0] == "alignment"
+
+    def test_match_by_list_embeddings_multiple_keywords(self, keyword_matcher):
+        """Test list embeddings matching with multiple keywords."""
+        keywords = ["sequence", "alignment", "protein"]
+        matches = keyword_matcher.match_by_list_embeddings(keywords, threshold=0.3)
+
+        assert isinstance(matches, list)
+        assert len(matches) > 0
+
+        # Check that each match has one of the keywords
+        for match in matches:
+            assert len(match.matched_keywords) == 1
+            assert match.matched_keywords[0] in keywords
+
+    def test_match_by_list_embeddings_threshold_filtering(self, keyword_matcher):
+        """Test that threshold properly filters results."""
+        keywords = ["alignment"]
+
+        # Lower threshold should return more results
+        matches_low = keyword_matcher.match_by_list_embeddings(keywords, threshold=0.2)
+        matches_high = keyword_matcher.match_by_list_embeddings(keywords, threshold=0.5)
+
+        assert len(matches_low) >= len(matches_high)
```

Comment:
I would also add an assert statement checking that all ther emaining matches have a higher confidence, and that none of the matches with higher confidence has been removed. Could check the exact expected length of matches.